src/order/well_founded_set.lean

```diff
@@ -707,3 +730,30 @@ begin
     refine ⟨g'.trans g, λ a b hab, (finset.forall_mem_cons _ _).2 _⟩,
     exact ⟨hg (order_hom_class.mono g' hab), hg' hab⟩ }
 end
+
+section prod_lex
+variables {rα : α → α → Prop} {rβ : β → β → Prop} {f : γ → α} {g : γ → β} {s : set γ}
+
+lemma well_founded.prod_lex (hα : well_founded (rα on f))
```

Comment:
I think this could do with a docstring comparing it to `prod.lex_wf`.

Some other questions:
* Should this name contain the word `on` somewhere?
* Does this generalize naturally to `sigma`? Indeed, your implementation seems to be using primarily sigma lemmas.
* Is `prod.lex rα rβ on λ c, (f c, g c)` the same as `prod.lex (rα on f) (rβ on g)`? I'm assuming there's a subtle difference, but that would likely be worth mentioning in a docstring.

Reply:
* Added a docstring
* Renamed to include `well_founded_on` and `fiber`
* I could get the statements to compile, but not the proofs. The indexing by `f` makes reasoning with `sigma` painful
```
section sigma_lex
variables {rι : ι → ι → Prop} {rπ : Π i, π i → π i → Prop} {f : γ → ι} {g : Π i, γ → π i}
  {s : set γ}

/-- Stronger version of `prod.lex_wf`. Instead of requiring `rβ on g` to be well-founded, we only
require it to be well-founded on fibers of `f`.-/
lemma well_founded.sigma_lex_of_well_founded_on_fiber (hι : well_founded (rι on f))
  (hπ : ∀ i, (f ⁻¹' {i}).well_founded_on (rπ i on g i)) :
  well_founded (sigma.lex rι rπ on λ c, ⟨f c, g (f c) c⟩) :=
begin
  refine ((psigma.lex_wf (well_founded_on_range.2 hι) $ λ a, hπ a).on_fun).mono (λ c c' h, _),
  exact λ c, ⟨⟨_, c, rfl⟩, c, rfl⟩,
  obtain h' | ⟨h', h''⟩ := sigma.lex_iff.1 h,
  { exact psigma.lex.left _ _ h' },
  { dsimp only [inv_image, (on)] at h' ⊢,
    convert psigma.lex.right (⟨_, c', rfl⟩ : range f) _ using 1, swap, -- stuck in he(q)ll
    exacts [⟨c, h'.1⟩, psigma.subtype_ext (subtype.ext h'.1) rfl, h'.2] }
end

lemma set.well_founded_on.prod_lex_of_well_founded_on_fiber (hι : s.well_founded_on (rι on f))
  (hπ : ∀ i, (s ∩ f ⁻¹' {i}).well_founded_on (rπ i on g i)) :
  s.well_founded_on (sigma.lex rι rπ on λ c, ⟨f c, g (f c) c⟩) :=
begin
  refine well_founded.sigma_lex_of_well_founded_on_fiber hι
    (λ i, subrelation.wf (λ b c h, _) (hπ i).on_fun), -- not happy either
  exact λ x, ⟨x, x.1.2, x.2⟩,
  assumption,
end

end sigma_lex
```
* `prod.lex rα rβ on λ c, (f c, g c) : γ → γ → Prop`, `prod.lex (rα on f) (rβ on g) : γ × γ → γ × γ → Prop`, so I'm not sure how to say there's a difference except that they don't even have the same type.

Reply:
The `he(q)ll` wasn't too bad. The second lemma just needed some unification help.